packages/SwingSet/src/vats/vattp/vat-vattp.js

```diff
@@ -16,181 +23,255 @@ import { Far } from '@endo/marshal';
+      connection: null,
+      connectionName,
+      exportedObjects: makeScalarBigMapStore(
+        `network host exported objects by address: ${connectionName} as ${allegedName}`,
+        { durable: true },
+      ),
+      nextAddressNumber: 1n,
+      // This assumes one connection per host.
+      // If that is not the case, receiver should instead be a
+      // durable makeScalarBigWeakMapStore keyed by connection.
+      receiver: null,
+      remoteLocatorUnum: null,
+    };
+  };
+  const makeNetworkHost = defineDurableKindMulti(
+    networkHostKindHandle,
+    initNetworkHostState,
+    {
+      host: {
+        publish: ({ state }, obj) => {
+          const address = `/alleged-name/${state.allegedName}/egress/${state.nextAddressNumber}`;
+          state.nextAddressNumber += 1n;
+          state.exportedObjects.init(address, obj);
+          return address;
+        },
+        lookup: ({ state }, address) => {
+          return E(state.remoteLocatorUnum).lookup(address);
+        },
+      },
+      commsSetReceiver: {
+        setReceiver: ({ state }, receiver) => {
+          assert(!state.receiver, X`setReceiver is call-once`);
+          assert(receiver, X`receiver must not be empty`);
+          state.receiver = receiver;
+        },
+      },
+      commsTransmitter: {
+        transmit: ({ state }, msg) => {
+          // 'msg' will be a string (vats/comms/outbound.js deliverToRemote)
+          E(state.connection).send(msg);
+        },
+      },
+      localLocatorUnum: {
+        lookup: ({ state }, address) => {
+          return state.exportedObjects.get(address);
+        },
+      },
+      handler: {
+        onOpen: async ({ state, facets: self }, connection, ..._args) => {
+          const name = state.connectionName;
+          assert(!state.connection, X`host ${name} already opened`);
+          state.connection = connection;
 
-  return handler;
+          const comms = state.comms;
+          // TODO: Can these calls be made in parallel?
+          await E(comms).addRemote(
+            name,
+            self.commsTransmitter,
+            self.commsSetReceiver,
+          );
+          await E(comms).addEgress(name, 0, self.localLocatorUnum);
+          state.remoteLocatorUnum = await E(comms).addIngress(name, 0);
+        },
+        onReceive: ({ state }, _connection, msg) => {
+          // setReceiver ought to be called before there's any chance of
+          // onReceive being called
+          E(state.receiver).receive(msg);
+        },
+        onClose: ({ state }, _connection, ..._args) => {
+          state.receiver = null;
+          // TODO: What would it take to fully support this?
+          console.warn(`deleting connection is not fully supported in comms`);
+        },
+        infoMessage: ({ state }, ...args) => {
+          void E(state.console).log('VatTP connection info:', ...args);
+        },
+      },
+    },
+  );
+
+  const vatFacets = vivifySingleton(baggage, 'vat-tp handler', {
```

Comment:
This value is an instance of a single-facet Kind, not a multi-Kind. So I'd use `vattpHandler` as the variable name instead of `vatFacets`, to avoid confusion.